src/Tools/dotnet-monitor/Exceptions/ExceptionsStoreLimitsCallback.cs

```
@@ -0,0 +1,104 @@
+        // Example: if exception A has an inner exception B, then A is an outer exception of B.
+
+        // Track the list of inner exceptions for a given exception; allows for quicker
+        // access to this relationship without having to look up IExceptionInstances.
+        private readonly Dictionary<ulong, List<ulong>> _innerExceptionMap = new();
+        // Track the list of outer exceptions for a given exception; allows for quicker
+        // access to this relationship without having to look up IExceptionInstances.
+        private readonly Dictionary<ulong, List<ulong>> _outerExceptionMap = new();
+        private readonly IExceptionsStore _store;
+        // Front of the list has the latest top-level exception ID and the
+        // back of the list has the oldest top-level exception ID.
+        private readonly LinkedList<ulong> _topLevelExceptions = new();
+        private readonly int _topLevelLimit;
+
+        public ExceptionsStoreLimitsCallback(IExceptionsStore store, int topLevelLimit)
+        {
+            ArgumentNullException.ThrowIfNull(store);
+#if NET8_0_OR_GREATER
+            ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(topLevelLimit, 0);
+#else
+            if (topLevelLimit <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(topLevelLimit));
+            }
+#endif
+
+            _store = store;
+            _topLevelLimit = topLevelLimit;
+        }
+
+        public override void BeforeAdd(IExceptionInstance instance)
+        {
+            _outerExceptionMap.Add(instance.Id, new List<ulong>());
+            _innerExceptionMap.Add(instance.Id, instance.InnerExceptionIds.ToList());
+
+            // Make sure all of the inner exceptions are no longer top-level exceptions.
+            foreach (ulong innerExceptionId in instance.InnerExceptionIds)
+            {
+                if (_outerExceptionMap.TryGetValue(innerExceptionId, out List<ulong> outerExceptions))
```

I know this code was just moved from an existing file, but I wanted to double check that we didn't overlook a scenario in the current version. If `_outerExceptionMap` doesn't contain the key of `innerExceptionId`, don't we still want to set `_outerExceptionMap[innerExceptionId] = new() { instance.Id }`?

---

That would mean that an inner exception was not reported previously because each exception gets added to the `_outerExceptionMap` at the begin of `BeforeAdd`, but this should not be possible unless events are dropped or exceptions are filtered at the collection site. I can take a look at mocking up this scenario in a separate change and addressing it there.

---

/todo Exception store callbacks should gracefully handle inner exceptions that were not reported